### PR TITLE
[ValueTracking] Implement `isKnownNonZero`/`computeKnownBits` for `llvm.vector.reduce.{add,mul}`

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[ValueTracking] Add tests for `computeKnownBits` of `llvm.vector.reduce.{add,mul}`; NFC**
- **[ValueTracking] Implement `computeKnownBits` for `llvm.vector.reduce.{add,mul}`**
- **[ValueTracking] Add tests for `isKnownNonZero` of `llvm.vector.reduce.{add,mul}`; NFC**
- **[ValueTracking] Implement `isKnownNonZero` for `llvm.vector.reduce.{add,mul}`**

### CODE REVIEW

**Comment by goldsteinn** [comment]
Most likely going nowhere. Just wanted to preserve in case it ever becomes desirable.

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 1]
```diff
@@ -2904,6 +2924,41 @@ static bool isKnownNonZeroFromOperator(const Operator *I,
       case Intrinsic::vector_reduce_smax:
       case Intrinsic::vector_reduce_smin:
         return isKnownNonZero(II->getArgOperand(0), Depth, Q);
+        // If we know the reduction doesn't overflow and all elements are
+        // non-zero, the reduction is non-zero.
+      case Intrinsic::vector_reduce_mul:
+      case Intrinsic::vector_reduce_add:
+        if (computeKnownBits(I, Depth + 1, Q).isNonZero())
+          return true;
```
Comment (dtcxzyw) [inline llvm/lib/Analysis/ValueTracking.cpp:2932]: It is incorrect. `llvm.vector.reduce.mul.i8(<4 x i8> <i8 4, i8 4, i8 4, i8 4>)` evaluates to zero.

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Analysis/ValueTracking.cpp (hunk 1 above)
Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp:2932]: This is `I`, the result, not the input vector.

Comment (dtcxzyw) [reply llvm/lib/Analysis/ValueTracking.cpp:2932]: Oh, sorry for my mistake.

Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp:2932]: BTW, I wouldn't spend your time reviewing this unless we decide to actually try to get it in.

**Pull request closed without merging**